### PR TITLE
Proposal for restart and init triggers

### PR DESCRIPTION
**Pull request by ahopkins**
WIP

### CODE REVIEW

File: jumpstarter/states.py [hunk 1]
```diff
@@ -39,22 +55,42 @@ def __init__(self, actor_state=ActorState):
             send_event=True,
         )
 
+        # TODO: Does transitions have an API for Enum based parallel?
```
Comment (thedrow) [inline jumpstarter/states.py:58]: It currently doesn't, unfortunately. If you can come up with an API, we can propose it to the transitions project.

**Review by thedrow — CHANGES_REQUESTED** [review]
There should be a callback after transitioning to the `started` that checks if we are in the `restarting` state and if we are, we should exit it.

File: jumpstarter/states.py [hunk 2]
```diff
@@ -19,17 +19,33 @@ class ActorStoppingState(Enum):
...
     stopped = auto()
+    restarting = ActorRestartingState
     crashed = auto()
 
 
+class ActorStateTriggers(str, Enum):
+    def _generate_next_value_(name, *args):
+        return name.lower()
+
+    init = auto()
```
Comment (thedrow) [inline jumpstarter/states.py:42]:
```suggestion
    initialize = auto()
```

I prefer to be clearer even if we're more verbose as we have auto-completion.

File: jumpstarter/states.py (hunk 1 above)
Comment (ahopkins) [reply jumpstarter/states.py]: Would be happy to.

**Comment by ahopkins** [comment]
@thedrow Do we want to go through the regular `starting` process again?

```
>>> my_actor.restart()
INFO:transitions.extensions.asyncio:Finished processing state started exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state restarting enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state restarting↦stopping enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state restarting↦stopping exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state restarting↦starting enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state restarting↦starting exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state restarting exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting↦dependencies_started enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting↦dependencies_started exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting↦resources_acquired enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting↦resources_acquired exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting↦tasks_started enter callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting↦tasks_started exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state starting exit callbacks.
INFO:transitions.extensions.asyncio:Finished processing state started enter callbacks.
> workflow.state=<ActorState.started: 3>
```

Or.... 
- restarting↦stopping
- restarting↦starting
- started

**Comment by ahopkins** [comment]
Also... I was playing with parallel states. I think using `children` with an initial value makes sense so that the callbacks can be ordered if needed.

**Comment by thedrow** [comment]
> @thedrow Do we want to go through the regular `starting` process again?

Yes we need to go through the same state transitions for both the shutdown process and the bootup process.
This allows us to release all resources, stop all the dependencies and tasks and start them again.

**Comment by thedrow** [comment]
> Also... I was playing with parallel states. I think using `children` with an initial value makes sense so that the callbacks can be ordered if needed.

If you can produce a state machine that does what you think is right, produce a diagram by inheriting from `HierarchicalAnyIOGraphMachine` instead of `HierarchicalAnyIOMachine` and paste it here.
Please refer to transitions' [documentation](https://github.com/pytransitions/transitions#diagrams) for details.

File: jumpstarter/states.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+class ActorStateMachine(HierarchicalAnyIOMachine):
+    def __init__(self, actor_state=ActorState):
+        super().__init__(
+            states=actor_state,
+            initial=actor_state.initializing,
+            auto_transitions=False,
+            send_event=True,
+        )
+
+        # TODO: Is there a way to do this with enums?
```
Comment (thedrow) [inline jumpstarter/states.py:58]: Yes, there is.
You need to nest enums.
See states.py.

Comment (ahopkins) [reply jumpstarter/states.py:58]: Nesting is not the issue. Providing additional params like `initial`. The current working solution I have relies upon that.

**Comment by thedrow** [comment]
Fixed by #28.

**Pull request closed without merging**